File: sesa-backend/src/main/java/com/sesa/salud/service/ReporteService.java
/**
 * Autor: Ing. J Sebastian Vargas S
 */
package com.sesa.salud.service;

import com.sesa.salud.dto.DashboardCitasPorDiaDto;
import com.sesa.salud.dto.DashboardCitasPorEstadoDto;
import com.sesa.salud.dto.DashboardFacturacionMesDto;
import com.sesa.salud.dto.DashboardMesCountDto;
import com.sesa.salud.dto.DashboardStatsDto;
import com.sesa.salud.dto.CumplimientoNormativoDto;
import com.sesa.salud.dto.IndicadorCalidadDto;
import com.sesa.salud.dto.ReporteResumenDto;
import com.sesa.salud.dto.PacienteRiesgoDto;
import com.sesa.salud.dto.RecuperacionCarteraDto;
import com.sesa.salud.dto.EvaluacionHcDto;
import com.sesa.salud.dto.AuditoriaHcProfesionalDto;
import com.sesa.salud.dto.AuditoriaHcServicioDto;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;

public interface ReporteService {
    ReporteResumenDto resumen();
    List<DashboardCitasPorDiaDto> citasPorUltimosDias(int dias);
    List<DashboardMesCountDto> consultasPorUltimosMeses(int meses);
    List<DashboardFacturacionMesDto> facturacionPorUltimosMeses(int meses);
    List<DashboardCitasPorEstadoDto> citasPorEstado();
    /** Estadísticas para gráficas del dashboard en una sola llamada (GET /api/reportes/dashboard). */
    DashboardStatsDto dashboardStats();

    /** Indicadores de calidad Res. 0256/2016 (Sistema de Información para la Calidad en Salud). */
    List<IndicadorCalidadDto> getIndicadoresCalidadRes256();

    /** Score de riesgo del paciente para cabecera HC (S1). */
    PacienteRiesgoDto getRiesgoPaciente(Long pacienteId);

    /** S4: Panel de cumplimiento normativo por período y opcionalmente por profesional. */
    CumplimientoNormativoDto getCumplimientoNormativo(LocalDate desde, LocalDate hasta, Long profesionalId);

    /** S9: Recuperación de cartera — glosas por período (y opcional contrato). */
    RecuperacionCarteraDto recuperacionCartera(Instant desde, Instant hasta, Long contratoId);

    /** S16: Evaluación de calidad de una consulta (motivo, CIE-10, plan, etc.). */
    EvaluacionHcDto evaluarAtencion(Long consultaId);

    /** S16: Reporte de auditoría HC por profesional (porcentaje completas, puntuación media). */
    List<AuditoriaHcProfesionalDto> reporteAuditoriaHcPorProfesional(LocalDate desde, LocalDate hasta);

    /** S16: Reporte de auditoría HC por servicio (tipo de consulta). */
    List<AuditoriaHcServicioDto> reporteAuditoriaHcPorServicio(LocalDate desde, LocalDate hasta);
}

File: sesa-salud/src/app/core/services/reporte.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface ReporteResumenDto {
  totalPacientes: number;
  totalCitas: number;
  totalConsultas: number;
  totalOrdenes: number;
  totalFacturas: number;
  totalFacturado: number;
}

export interface DashboardCitasPorDiaDto {
  fecha: string;
  total: number;
}

export interface DashboardMesCountDto {
  anio: number;
  mes: number;
  total: number;
}

export interface DashboardFacturacionMesDto {
  anio: number;
  mes: number;
  cantidad: number;
  valorTotal: number;
}

export interface DashboardCitasPorEstadoDto {
  estado: string;
  total: number;
}

/** Respuesta del endpoint único GET /api/reportes/dashboard */
export interface DashboardStatsDto {
  citasPorDia?: DashboardCitasPorDiaDto[];
  consultasPorMes?: DashboardMesCountDto[];
  facturacionPorMes?: DashboardFacturacionMesDto[];
  citasPorEstado?: DashboardCitasPorEstadoDto[];
}

/** Indicador de calidad en salud — Res. 0256/2016 */
export interface IndicadorCalidadDto {
  codigo: string;
  nombre: string;
  categoria: string;
  valor: string;
  meta: string;
  unidad: string;
  interpretacion: string;
}

/** Score de riesgo del paciente para cabecera HC (S1). */
export interface PacienteRiesgoDto {
  nivelRiesgo: string;
  puntos: number;
  factores: string[];
  recomendaciones: string[];
}

/** S4: Panel de cumplimiento normativo. */
export interface CumplimientoNormativoDto {
  periodoInicio: string;
  periodoFin: string;
  profesionalId?: number;
  porcentajeRdaEnviado: number;
  totalAtenciones: number;
  atencionesConRdaEnviado: number;
  porcentajeUrgenciasEnTiempo: number;
  totalUrgenciasAtendidas: number;
  urgenciasDentroTiempo: number;
  porcentajeHcConCie10YEvolucion24h: number;
  atencionesConCie10YEvolucion24h: number;
  totalResultadosCriticosNoLeidos: number;
  indicadores0256: IndicadorCalidadDto[];
}

/** S16: Auditoría de calidad HC — evaluación por atención. */
export interface EvaluacionHcDto {
  consultaId: number;
  camposCompletos: boolean;
  camposFaltantes: string[];
  puntuacion: number;
  umbralAceptable: number;
}

/** S16: Resumen por profesional. */
export interface AuditoriaHcProfesionalDto {
  profesionalId: number;
  profesionalNombre: string;
  totalAtenciones: number;
  atencionesCompletas: number;
  porcentajeCompletas: number;
  puntuacionMedia: number;
  bajoUmbral: boolean;
}

/** S16: Resumen por servicio (tipo consulta). */
export interface AuditoriaHcServicioDto {
  servicioId: number;
  servicioNombre: string;
  totalAtenciones: number;
  atencionesCompletas: number;
  porcentajeCompletas: number;
  puntuacionMedia: number;
  bajoUmbral: boolean;
}

@Injectable({ providedIn: 'root' })
export class ReporteService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.apiUrl}/reportes`;

  resumen(): Observable<ReporteResumenDto> {
    return this.http.get<ReporteResumenDto>(`${this.apiUrl}/resumen`);
  }

  /**
   * Una sola llamada para todas las gráficas del dashboard.
   * URL: GET /api/reportes/dashboard (el backend tiene context-path /api).
   */
  dashboardStats(): Observable<DashboardStatsDto> {
    return this.http.get<DashboardStatsDto>(`${this.apiUrl}/dashboard`);
  }

  citasPorDia(dias: number = 7): Observable<DashboardCitasPorDiaDto[]> {
    const params = new HttpParams().set('dias', String(dias));
    return this.http.get<DashboardCitasPorDiaDto[]>(`${this.apiUrl}/dashboard/citas-por-dia`, { params });
  }

  consultasPorMes(meses: number = 6): Observable<DashboardMesCountDto[]> {
    const params = new HttpParams().set('meses', String(meses));
    return this.http.get<DashboardMesCountDto[]>(`${this.apiUrl}/dashboard/consultas-por-mes`, { params });
  }

  facturacionPorMes(meses: number = 6): Observable<DashboardFacturacionMesDto[]> {
    const params = new HttpParams().set('meses', String(meses));
    return this.http.get<DashboardFacturacionMesDto[]>(`${this.apiUrl}/dashboard/facturacion-por-mes`, { params });
  }

  citasPorEstado(): Observable<DashboardCitasPorEstadoDto[]> {
    return this.http.get<DashboardCitasPorEstadoDto[]>(`${this.apiUrl}/dashboard/citas-por-estado`);
  }

  /** Tablero de calidad en salud — Res. 0256/2016 */
  indicadoresCalidad(): Observable<IndicadorCalidadDto[]> {
    return this.http.get<IndicadorCalidadDto[]>(`${this.apiUrl}/calidad`);
  }

  /** Score de riesgo del paciente para cabecera de historia clínica (S1). */
  getRiesgoPaciente(pacienteId: number): Observable<PacienteRiesgoDto> {
    return this.http.get<PacienteRiesgoDto>(`${this.apiUrl}/paciente/${pacienteId}/riesgo`);
  }

  /** S4: Panel de cumplimiento normativo. */
  getCumplimientoNormativo(params: {
    desde?: string;
    hasta?: string;
    profesionalId?: number;
  }): Observable<CumplimientoNormativoDto> {
    let httpParams = new HttpParams();
    if (params.desde) httpParams = httpParams.set('desde', params.desde);
    if (params.hasta) httpParams = httpParams.set('hasta', params.hasta);
    if (params.profesionalId != null) httpParams = httpParams.set('profesionalId', String(params.profesionalId));
    return this.http.get<CumplimientoNormativoDto>(`${this.apiUrl}/cumplimiento-normativo`, { params: httpParams });
  }

  /** S16: Auditoría de calidad HC por profesional. */
  getAuditoriaHcPorProfesional(params: { desde?: string; hasta?: string }): Observable<AuditoriaHcProfesionalDto[]> {
    let httpParams = new HttpParams();
    if (params.desde) httpParams = httpParams.set('desde', params.desde);
    if (params.hasta) httpParams = httpParams.set('hasta', params.hasta);
    return this.http.get<AuditoriaHcProfesionalDto[]>(`${this.apiUrl}/auditoria-hc/por-profesional`, { params: httpParams });
  }

  /** S16: Auditoría de calidad HC por servicio. */
  getAuditoriaHcPorServicio(params: { desde?: string; hasta?: string }): Observable<AuditoriaHcServicioDto[]> {
    let httpParams = new HttpParams();
    if (params.desde) httpParams = httpParams.set('desde', params.desde);
    if (params.hasta) httpParams = httpParams.set('hasta', params.hasta);
    return this.http.get<AuditoriaHcServicioDto[]>(`${this.apiUrl}/auditoria-hc/por-servicio`, { params: httpParams });
  }
}
